### PR TITLE
Sortedcontainerupdates

### PR DESCRIPTION
**Pull request by StephenVavasis**
Major update to sorted containers.

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/239: Dict constructors on typed data but without explicit type declarations should no longer return `{Any,Any}` items because collect is now used when explicit types are omitted

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/479: Deprecate `insert!`, `pop!(s)`, introduce `popfirst!`, `poplast!`, `ss_push!`, `sd_push!`, `smd_push!`

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/528: `push!` documentation improved

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/667: Docs extensively cleaned up and much greater use of documenter.jl

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/669: Expanded use of tokens including `a:b` notation for tokens

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/671: Add reverse iteration

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/742: New union, merge, symdiff, setdiff, intersect algorithms for sorted containers to improve performance

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/726: Clarify docs on `ss_push!`, `sd_push!`, and `sdm_push!` (formerly `insert!`)

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/767: Implement firstindex, lastindex, reverse for sorted container iteration

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/781: Use collect on an intermediate step
to infer types in untyped construction of sorted containers

Also, this commit cleans up the constructors and adds two new iteration-helper functions `inclusive_key` and `exclusive_key` that were useful in my own work.

### CODE REVIEW

**Comment by StephenVavasis** [comment]
The following test from the new testset, which is meant to exercise a new `show` method, is failing on some platforms:
```julia
    my_assert(remove_spaces(repr(MIME{Symbol("text/plain")}(), factors5)) ==
              "SortedMultiDict{Char,Int64,ForwardOrdering}" *
              "(ForwardOrdering(),['a'=>1,'b'=>2,'c'=>3])")
```
Here, `remove_spaces` is a locally written function-- see below.  So apparently on some platforms, `show` works differently?  Anyone know the differences?  Or is there a better way to write a unit test for `show`?

```julia
function remove_spaces(s::String)
    b = Vector{UInt8}()
    for c in s
        if !isspace(c)
            push!(b,UInt8(c))
        end
    end
    String(b)
end
```

**Comment by StephenVavasis** [comment]
The reason the test is failing is: In Julia 1.0, printing the vector `[1=>'a',2=>'b']` prepends the output with `Pair{Int64,Char}`.  Prepending this eltype-declaration does not happen in Julia 1.7.  Apparently I need to think of a more robust unit test for a `show` method.

**Comment by StephenVavasis** [comment]
I replaced the failing unit test for `show` that does not require precisely a particular string, and now all tests are successful.

**Review by oxinabox — COMMENTED** [review]
sorry I have not finished reviewing this.
This broadly speaking looks good and if all the existing tests still pass then it should be good

File: src/sorted_container_iteration.jl
Comment (StephenVavasis) [reply src/sorted_container_iteration.jl:16]: Are you asking why `Token` is implemented as a tuple rather than a `struct`?   This is a legacy issue.  In pre-1.0 Julia, any struct that contained an immutable object had itself to be heap allocated.  However, in pre-1.0 Julia, operations on tuples containing an immutable object didn't always allocate.  In particular, `deref((mysorteddict, semitok))` did not heap-allocate its 2-tuple argument, whereas a hypothetical `deref(Token(mysorteddict, semitok))`, where `Token` is the constructor of a hypothetical struct, would have allocated.

File: src/DataStructures.jl
Comment (oxinabox) [inline src/DataStructures.jl:73]: Rather than all 3 `sd_push!, ss_push!, smd_push!` is there a single name we can overload for these that refers to what it returns, rather than hungarian notation for it's first argument?
Like `push_and_get_token!` or something? (is that what they are doing?)

File: src/sorted_container_iteration.jl
Comment (oxinabox) [reply src/sorted_container_iteration.jl:16]: More my surprise was that a token included both the container and the indexing object.
I would have expected the token to just be the indexing object (but that is defined as a semitoken)
is the idea that the token is enough to fully retrieve the stored object?
But semitoken is only half of what is expected?

File: src/sorted_container_iteration.jl
Comment (oxinabox) [inline src/sorted_container_iteration.jl:45]: do we need this to be exported?
It seems like if someone needs it is is a trivial and obvious definition.

File: src/sorted_container_iteration.jl
Comment (oxinabox) [inline src/sorted_container_iteration.jl:58]: as per `startof` do we need this? or can we just use `lastindex`?

File: src/sorted_container_iteration.jl
Comment (oxinabox) [inline src/sorted_container_iteration.jl:68]: as per `token_firstindex` do we need to export this?
It is a trivial and obvious definition

File: src/sorted_container_iteration.jl
Comment (oxinabox) [reply src/sorted_container_iteration.jl:254]: similar for the other `foo`/`foo_nocheck` pairs?

File: src/sorted_container_iteration.jl
Comment (oxinabox) [inline src/sorted_container_iteration.jl:1057]: should we comment here that if `isequal` is defined consistently with `isless` etc this should have same behavour?

File: src/sorted_container_iteration.jl
Comment (oxinabox) [inline src/sorted_container_iteration.jl:1120]: why not define this as `:` ?

File: src/sorted_container_iteration.jl
Comment (oxinabox) [inline src/sorted_container_iteration.jl:1141]: this easily fits on 1 line, no?
```suggestion
+(t1::Token, numstep::Integer) = numstep >= 0 ? stepforward(t1, numstep) : stepback(t1, -numstep)

```

File: src/sorted_container_iteration.jl
Comment (oxinabox) [inline src/sorted_container_iteration.jl:1143]: this easily fits on 1 line, no?
```suggestion
-(t1::Token, numstep::Integer) = numstep >= 0 ? stepback(t1, numstep) : stepforward(t1, -numstep)
```

File: src/sorted_container_iteration.jl
Comment (oxinabox) [inline src/sorted_container_iteration.jl:1149]: what do you think of
```suggestion
    (j in m.bt.useddatacells) || throw(BoundsError())
```

File: src/sorted_dict.jl
Comment (oxinabox) [inline src/sorted_dict.jl:230]: can we move this into `src/deprecations.jl` so we remember to delete it in next release?

File: test/test_sorted_containers.jl
Comment (oxinabox) [inline test/test_sorted_containers.jl:2]: this fits on one line
```suggestion
using Base: Ordering, Forward, Reverse, ForwardOrdering, ReverseOrdering
```

File: test/test_sorted_containers.jl
Comment (oxinabox) [inline test/test_sorted_containers.jl:5]: We should make the imports into `using`s i think

File: test/test_sorted_containers.jl
Comment (oxinabox) [inline test/test_sorted_containers.jl:4]: isn't this called `Base.lessthan` now days?

File: test/test_sorted_containers.jl
Comment (oxinabox) [inline test/test_sorted_containers.jl:14]:
```suggestion
# for testing custom ordering
```

File: test/test_sorted_containers.jl
Comment (oxinabox) [inline test/test_sorted_containers.jl:50]:
```suggestion
remove_spaces(s::String) = replace(s, re"\s+"=>"")    
```

File: test/test_sorted_containers.jl
Comment (oxinabox) [inline test/test_sorted_containers.jl:2096]: can we label the new tests with a comment including the URL of the github issues that they prove is solved?

**Review by oxinabox — COMMENTED** [review]
Nice work.
I assume that the logic is at least as correct as before as all previous tests still pass.

Some mostly stylistic comments to resolve

File: src/DataStructures.jl
Comment (StephenVavasis) [reply src/DataStructures.jl:72]: I am having trouble with this for `==` and `:`.  In particular, the following constructs give me syntax errors:
```
Base.(==)(a,b) = ...
Base.==(a,b) = ...
function Base.(:)(a,b)
```
whereas there are no syntax errors with
```
import Base.==
import Base.(:)

==(a,b) = ...
function (:)(a,b)
```
Any thoughts how to fix this?

File: src/sorted_container_iteration.jl
Comment (StephenVavasis) [reply src/sorted_container_iteration.jl:1057]: The issue here is not whether `isless` and `isequal` are consistent.  Rather, the issue is whether the user has defined an ordering (such as case-insensitive for strings) for the container different from `isequal` and `isless`.

File: src/sorted_container_iteration.jl
Comment (StephenVavasis) [reply src/sorted_container_iteration.jl:1120]: The definition `function Base.(:)(a,b)` gives a syntax error (noted above).  However, `function (:)(a,b)` works, so I have fixed this (but I still have the issue with `import` noted above).

File: src/sorted_container_iteration.jl
Comment (StephenVavasis) [reply src/sorted_container_iteration.jl:1149]: I can make this change, but I have used the idiom `!(correct) && throw(...)` (rather than `correct || throw(...)`) throughout my code.  Do you recommend that I change them all.

File: test/test_sorted_containers.jl
Comment (StephenVavasis) [reply test/test_sorted_containers.jl:4]: I checked `ordering.jl` in `Base`, and I still see `lt`.  Where did you see `lessthan`?

File: src/sorted_container_iteration.jl
Comment (StephenVavasis) [reply src/sorted_container_iteration.jl:1120]: OK, I figured this out.

File: src/DataStructures.jl
Comment (StephenVavasis) [reply src/DataStructures.jl:72]: I figured this out.

File: src/sorted_container_iteration.jl
Comment (StephenVavasis) [reply src/sorted_container_iteration.jl:254]: I tried to implement this, but it is not working properly, so perhaps I am misunderstanding the rules.  Here is the new definition of `deref`:

```julia
Base.@propagate_inbounds function deref(ii::SortedDictToken)
    Base.@boundscheck has_data(ii)
    @inbounds kdrec = ii[1].bt.data[ii[2].address]
    return Pair(kdrec.k, kdrec.d)
end
```

My intention is that if a `deref` operation is prefixed by `@inbounds`, then `has_data` will not be invoked.  But this is not happening:

```julia
julia> s = SortedDict{Int,String}()
SortedDict{Int64, String, Base.Order.ForwardOrdering}()

julia> deref(token_firstindex(s))
ERROR: BoundsError
Stacktrace:
 [1] has_data
   @ c:\Users\vavasis\OneDrive - University of Waterloo\Julia_pkg_devdir\DataStructures.jl\src\sorted_container_iteration.jl:442 [inlined]
 [2] deref(ii::Tuple{SortedDict{Int64, String, Base.Order.ForwardOrdering}, DataStructures.Tokens.IntSemiToken})
   @ DataStructures c:\Users\vavasis\OneDrive - University of Waterloo\Julia_pkg_devdir\DataStructures.jl\src\sorted_container_iteration.jl:242
 [3] top-level scope
   @ REPL[24]:1

julia> @inbounds deref(token_firstindex(s))
ERROR: BoundsError
Stacktrace:
 [1] has_data
   @ c:\Users\vavasis\OneDrive - University of Waterloo\Julia_pkg_devdir\DataStructures.jl\src\sorted_container_iteration.jl:442 [inlined]
 [2] deref(ii::Tuple{SortedDict{Int64, String, Base.Order.ForwardOrdering}, DataStructures.Tokens.IntSemiToken})
   @ DataStructures c:\Users\vavasis\OneDrive - University of Waterloo\Julia_pkg_devdir\DataStructures.jl\src\sorted_container_iteration.jl:242
 [3] top-level scope
   @ REPL[25]:1
```

Comment (StephenVavasis) [reply src/sorted_container_iteration.jl:254]: OK, nvm, I figured this out too.  Apparently `@inbounds` is not necessarily active if it is called directly from the REPL as opposed to its being called from a compiled function.

File: src/sorted_container_iteration.jl
Comment (StephenVavasis) [reply src/sorted_container_iteration.jl:41]: I would like to leave these.  See other comment.

File: src/sorted_container_iteration.jl
Comment (StephenVavasis) [reply src/sorted_container_iteration.jl:16]: I think full tokens are useful, e.g., `deref(token_firstindex(s))` as opposed to `deref((s,firstindex(s)))` where the user needs to be consistent with the argument `s`.  So if it is OK with you, I would like to leave these.

File: src/sorted_container_iteration.jl
Comment (StephenVavasis) [reply src/sorted_container_iteration.jl:412]: I didn't know this!  However, if you don't mind, I'll leave it as it is because there is not much benefit in changing it compared to the effort needed.

File: docs/Project.toml
Comment (StephenVavasis) [reply docs/Project.toml:5]: OK, I put it back.  I'm not sure how it got deleted.

File: docs/src/sorted_containers.md
Comment (StephenVavasis) [reply docs/src/sorted_containers.md:25]: done

File: docs/src/sorted_containers.md
Comment (StephenVavasis) [reply docs/src/sorted_containers.md:72]: done

File: docs/src/sorted_containers.md
Comment (StephenVavasis) [reply docs/src/sorted_containers.md:198]: I didn't know about the other syntax for documenter, but now that you told me, I have fixed it.

File: docs/src/sorted_containers.md
Comment (StephenVavasis) [reply docs/src/sorted_containers.md:271]: Done

File: src/DataStructures.jl
Comment (StephenVavasis) [reply src/DataStructures.jl:73]: Done.  The new function is called `push_return_token!`.  Since it is a single function name, I unified the APIs by making the return variables `Bool,IntSemiToken`. The `Bool` return variable is redundant for SortedMultiDict because push always creates a new record for SortedMultiDict (i.e., the first return variable is always `true` for a SortedMultiDict).

File: src/sorted_container_iteration.jl
Comment (StephenVavasis) [reply src/sorted_container_iteration.jl:58]: I have deprecated `startof` and `endof` in favor of `firstindex` and `lastindex`.

File: src/sorted_container_iteration.jl
Comment (StephenVavasis) [reply src/sorted_container_iteration.jl:45]: See other comment about `token_firstindex`.

File: src/sorted_container_iteration.jl
Comment (StephenVavasis) [reply src/sorted_container_iteration.jl:68]: See other comment about `token_firstindex`

File: src/sorted_container_iteration.jl
Comment (StephenVavasis) [reply src/sorted_container_iteration.jl:1141]: Done.

File: src/sorted_container_iteration.jl
Comment (StephenVavasis) [reply src/sorted_container_iteration.jl:1143]: Done.

File: src/sorted_dict.jl
Comment (StephenVavasis) [reply src/sorted_dict.jl:230]: Done

File: test/test_sorted_containers.jl
Comment (StephenVavasis) [reply test/test_sorted_containers.jl:2]: Done

File: test/test_sorted_containers.jl
Comment (StephenVavasis) [reply test/test_sorted_containers.jl:5]: Done

File: test/test_sorted_containers.jl
Comment (StephenVavasis) [reply test/test_sorted_containers.jl:50]: I found out that it is actually `r"\s+"` rather than `re"\s+"`, but aside from that, this is done.

**Comment by StephenVavasis** [comment]
@oxinabox I believe I have addressed all of your comments in this PR.  In the last commit that I just made, I have added comments to the test sets to indicate where the various issues are tested.  Let me know if you need anything else before you can merge this PR.

File: src/DataStructures.jl
Comment (oxinabox) [inline src/DataStructures.jl:45]: We don't need to explictly export deprecated things `@deprecate` does that for us

```suggestion
```

**Review by oxinabox — APPROVED** [review]
Sounds good.
You have merge rights.
Merge when you are happy

**Pull request merged**
